static/app/components/tabs/tabList.tsx

```diff
@@ -0,0 +1,213 @@
+import {useContext, useEffect, useMemo, useRef, useState} from 'react';
+import styled from '@emotion/styled';
+import {AriaTabListProps, useTabList} from '@react-aria/tabs';
+import {TabListProps as TabListStateProps, useTabListState} from '@react-stately/tabs';
+import {Orientation} from '@react-types/shared';
+
+import DropdownButton from 'sentry/components/dropdownButton';
+import CompactSelect from 'sentry/components/forms/compactSelect';
+import {IconEllipsis} from 'sentry/icons';
+import {t} from 'sentry/locale';
+import space from 'sentry/styles/space';
+
+import {TabsContext} from './index';
+import {Tab} from './tab';
+
+interface TabListProps<T> extends TabListStateProps<T>, AriaTabListProps<T> {
+  className?: string;
+}
+
+/**
+ * Uses IntersectionObserver API to detect overflowing tabs. Returns an array
+ * containing of keys of overflowing tabs.
+ */
+function useOverflowTabs({
+  tabList,
+  tabs,
+}: {
+  tabList: HTMLUListElement | null;
+  tabs: Record<React.Key, HTMLLIElement | null>;
+}) {
+  const [overflowTabs, setOverflowTabs] = useState<React.Key[]>([]);
+
+  useEffect(() => {
+    const options = {
+      root: tabList,
+      // Nagative right margin to account for overflow menu's trigger button
+      rootMargin: `0px -42px 0px ${space(1)}`,
+      threshold: 1,
+    };
+
+    const callback: IntersectionObserverCallback = entries => {
+      entries.forEach(entry => {
+        const {target} = entry;
+        const {key} = (target as HTMLElement).dataset;
+        if (!key) {
+          return;
+        }
+
+        if (!entry.isIntersecting) {
+          setOverflowTabs(prev => prev.concat([key]));
+          return;
+        }
+
+        setOverflowTabs(prev => prev.filter(k => k !== key));
+      });
+    };
+
+    const observer = new IntersectionObserver(callback, options);
+    Object.values(tabs ?? {}).forEach(element => element && observer.observe(element));
+
+    return () => observer.disconnect();
+    // Suppress an exhaustive-deps warning about adding `tabListRef` and
```

Comment:
I don't see those used in here? Is this outdated?

Reply:
Ah yes, it's `tabList` and `tabs`, not `tabListRef` and `tabRefs`. Will update the comment.

Reply:
checkout my later comment. I think you should actually pass in the ref objects probably